### PR TITLE
Don't lint `let_unit_value` when needed for type inferenece

### PR DESCRIPTION
**Pull request by Jarcho**
fixes: #1502

Pinging @dtolnay. I think this is enough to fix the issue. Do you have a good list crates to test this on?

changelog: Don't lint `let_unit_value` when needed for type inference

### CODE REVIEW

File: tests/ui/let_unit.rs [hunk 1]
```diff
@@ -22,7 +22,7 @@ fn main() {
 
     multiline_sugg();
 
-    let_and_return!(()) // should be fine
+    let_and_return!(()) // should be fine}
```
Comment (llogiq) [inline tests/ui/let_unit.rs:25]: Is this intended?

**Review by llogiq — COMMENTED** [review]
I don't know which crates suffered this problem, but code-wise this looks good to me.

File: tests/ui/let_unit.rs (hunk 1 above)
Comment (Jarcho) [reply tests/ui/let_unit.rs]: Fixed.

**Comment by llogiq** [comment]
Sorry for letting this wait for so long, but I lack insight into actual cases in the wild, so I'm not sure if they are all caught 

However, as even if this misses some cases it is still a good improvement, r=me with a rebase.

**Comment by Jarcho** [comment]
This will now handle block expression and branches when checking if type inference is needed.

The only thing I can think of that would also be a false positive would be a something like:
```rust
trait Foo {
    type T;
    fn get_t(&self) -> Self::T;
    fn use_t(&self, _: T);
}
struct Bar;
impl Foo for Bar {
    type T = ();
    ...
}

fn f(bar: Bar) {
    let x = bar.get_t();
    bar.use_t(x);
}
```
But I'm pretty sure that already doesn't lint.

**Comment by llogiq** [comment]
Ok, I might be offline for a while, but r+ when rebased. Thanks and again sorry for the wait.

@bors delegate+

**Pull request merged**